nss/aad_auth/getent.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/ubuntu/aad-auth/internal/cache"
+	"github.com/ubuntu/aad-auth/internal/logger"
+	"github.com/ubuntu/aad-auth/internal/nss"
+	"github.com/ubuntu/aad-auth/internal/nss/group"
+	"github.com/ubuntu/aad-auth/internal/nss/passwd"
+	"github.com/ubuntu/aad-auth/internal/nss/shadow"
+)
+
+// Getent processes the args and queries the database for the requested entries.
+func Getent(ctx context.Context, dbName, key string, cacheOpts ...cache.Option) string {
+	logger.Debug(ctx, "Getting entry %q from %s ", key, dbName)
+
+	var entries []fmt.Stringer
+	var err error
+	if key != "" {
+		var e fmt.Stringer
+		e, err = getEntryByKey(ctx, dbName, key, cacheOpts...)
+		entries = []fmt.Stringer{e}
+		if err != nil {
+			entries = nil
+		}
+	} else {
+		entries, err = getAllEntries(ctx, dbName, cacheOpts...)
+	}
+
+	return fmtGetentOutput(ctx, entries, err)
+}
+
+func getEntryByKey(ctx context.Context, dbName, key string, cacheOpts ...cache.Option) (entry fmt.Stringer, err error) {
+	u, err := strconv.ParseUint(key, 10, 64)
+	if err != nil {
+		return getEntryByName(ctx, dbName, key, cacheOpts...)
+	}
+	return getEntryByID(ctx, dbName, uint(u), cacheOpts...)
+}
+
+func getEntryByName(ctx context.Context, dbName, name string, cacheOpts ...cache.Option) (entry fmt.Stringer, err error) {
+	logger.Debug(ctx, "Getting entry with name = %s from %s", name, dbName)
+
+	var e fmt.Stringer
+
+	switch dbName {
+	case "passwd":
+		e, err = passwd.NewByName(ctx, name, cacheOpts...)
+	case "group":
+		e, err = group.NewByName(ctx, name, cacheOpts...)
+	case "shadow":
+		e, err = shadow.NewByName(ctx, name, cacheOpts...)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+func getEntryByID(ctx context.Context, dbName string, id uint, cacheOpts ...cache.Option) (entry fmt.Stringer, err error) {
+	logger.Debug(ctx, "Getting entry with id = %d from %s", id, dbName)
+
+	var e fmt.Stringer
+
+	switch dbName {
+	case "passwd":
+		e, err = passwd.NewByUID(ctx, id, cacheOpts...)
+	case "group":
+		e, err = group.NewByGID(ctx, id, cacheOpts...)
+	case "shadow":
+		return nil, nss.ErrNotFoundENoEnt
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+func getAllEntries(ctx context.Context, dbName string, cacheOpts ...cache.Option) (entries []fmt.Stringer, err error) {
+	logger.Debug(ctx, "Getting all entries from %s", dbName)
+
+	start, end := initIterationForDB(dbName)
+	if err = start(ctx, cacheOpts...); err != nil {
+		return nil, err
+	}
+	defer end(ctx)
+
+	for {
+		var entry fmt.Stringer
+		entry, err = nextEntryForDB(ctx, dbName)
+		if err != nil {
+			// The iteration ends with ErrNoEnt, even when it's successful.
+			break
+		}
+		entries = append(entries, entry)
+	}
+
+	if !errors.Is(err, nss.ErrNotFoundENoEnt) {
```

Comment:
Indeed, it was a good idea to remove the defer() now that the function is simpler. It’s more readable, thanks!